[SPARK-2996] Implement userClassPathFirst for driver, yarn.

vanzin opened the pull request:
Yarn's config option `spark.yarn.user.classpath.first` does not work the same way as
`spark.files.userClassPathFirst`; Yarn's version is a lot more dangerous, in that it
modifies the system classpath, instead of restricting the changes to the user's class
loader. So this change implements the behavior of the latter for Yarn, and deprecates
the more dangerous choice.

To be able to achieve feature-parity, I also implemented the option for drivers (the existing
option only applies to executors). So now there are two options, each controlling whether
to apply userClassPathFirst to the driver or executors. The old option was deprecated, and
aliased to the new one (`spark.executor.userClassPathFirst`).

The existing "child-first" class loader also had to be fixed. It didn't handle resources, and it
was also doing some things that ended up causing JVM errors depending on how things
were being called.

Commit:
Add class path isolation support for Yarn cluster mode.

This change adds support for using ChildExecutorURLClassLoader
within Yarn, to give user classes preference over system classes
while still allowing Spark to load its own version of conflicting
classes. This is similar to the existing
"spark.files.userClassPathFirst" configuration that standalone
mode has.

To avoid breaking the semantics of the existing options, I added
two new options ("spark.{driver,executor}.userClassPathFirst"),
the second of which replaces the existing option, so that users
are not surprised that now their driver is also being affected
by the option.

One potentially breaking change is that ChildExecutorURLClassLoader
now also applies the same logic (user first) to resources, not
just classes; this should be fine since it's more in line with
the spirit of the option anyway.

Commit:
Re-work CoarseGrainedExecutorBackend command line arguments.

Preparation for changes to come.

Commit:
[yarn] Enable executor class path isolation.

This requires plumbing some initial state into the Executor instance,
since Yarn distributes some user files before the containers are
started, and those need to be added to the user's class path for
things to work.

Also cleaned up some other things in ClientBase to make the shared
code more generic.

Commit:
Enable driver class path isolation in SparkSubmit.

Make SparkSubmit respect the new option for driver isolation. This
covers client mode for, hopefully, all cluster managers.

Added simple SparkSubmit tests and also yarn-client test to make
sure changes work as expected.

Commit:
Update doc with new option, change name to "userClassPathFirst".

Matches the previous one and is actually more descriptive of what
it actually does.

Commit:
Add "deprecated configs" to SparkConf.

This allows code to just have to handle the current config names
instead of having to look for all the deprecated ones too.

Commit:
Fix the child-first class loader.

The class loader could run into a situation where a class managed by the
child class loader referenced a still-unloaded class from the parent
class loader. Since the child class loader ("userClassLoader" in the
previous code) did not have a parent, finding that class would fail,
and the child class would fail to load.

A second problem was calling `loadClass()` from a `findClass()`
implementation. This could cause the code to try to define the
same class more than once.

The fix for both is to override `loadClass()` instead of `findClass()`,
and implement the "priority inversion" there.

Commit:
Add deprecation warning to yarn option.

Also, only print warning when setting configuration values. And
use an atomic boolean to pretty up some ugly code.

Commit:
Run all Yarn integration tests via spark-submit.

spark-submit is needed for the class path tests, so just run
all tests using it instead of potentially polluting the current
VM and using internal APIs.

Also remove extra log info from child process since the thread name
already contains that information.

Commit:
Fix ClientBaseSuite.

Use the (deprecated) spark.yarn.user.classpath.first option to
force user jars to show up in the classpath.

Commit:
Register user jars in executor up front.

Avoid extra copying of files already distributed by Yarn, which could
lead to problems (Utils.fetchFile() trying to change permissions of
file managed by Yarn container). This required changing the code
to check for up-to-date jars slightly, so that it considers the file
name and not the HTTP URL sent by the driver.

Also rename a method to be consistent with other uses of the config
option.

Commit:
Fix yarn-cluster mode without userClassPathFirst.

Still need to instantiate the user class loader to load the
user jars.

Commit:
Fix Yarn executor command line.

Need to pass user jars in the command line regardless of the
userClassPathFirst configuration.

Commit:
Implement spark.driver.userClassPathFirst for standalone cluster mode.

Commit:
Cleanups.

Mostly minor, except for ClientBase.scala where I cleaned up the
generated classpath to only include the necessary entries.

Commit:
Remove write check on user jars.

Seems that I can't really hit the issue when my cluster's clocks
are properly synchronized, so remove the hack.

Commit:
Add test file to rat exclude.

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
	core/src/main/scala/org/apache/spark/executor/Executor.scala

Commit:
Merge branch 'master' into SPARK-2996

Commit:
Remove now outdated message.

Commit:
Merge branch 'master' into SPARK-2996

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	.rat-excludes
	core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
	core/src/main/scala/org/apache/spark/executor/Executor.scala
	docs/configuration.md
	yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientBase.scala

tgravescs commented:
sorry for the delay on the reviews.. to much going on with holidays and such. I hope to get to this early next week.

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientBase.scala
	yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnableUtil.scala
	yarn/src/test/resources/log4j.properties
	yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala

SparkQA commented:
[Test build #25268 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25268/consoleFull) for   PR 3233 at commit [`3730151`](https://github.com/apache/spark/commit/3730151df5f53921ed6840a78ac29ef3be0108d9).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25268 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25268/consoleFull) for   PR 3233 at commit [`3730151`](https://github.com/apache/spark/commit/3730151df5f53921ed6840a78ac29ef3be0108d9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            logInfo("Adding " + url + " to class loader")`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25268/
Test PASSed.

tgravescs commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -373,4 +388,62 @@ private[spark] object SparkConf {
+   * @param warn Whether to print a warning if the key is depreated. Warnings will be printed
+   *             only once for each key.
+   */
+  def getConfKey(userKey: String, warn: Boolean = false): String = {
```
warn flag isn't being used

andrewor14 replied:
It's a little weird for this to be `getConfKey(key)`. From the name it looks like an identity function. I think a better name is `translateConfKey`

tgravescs commented on core/src/main/scala/org/apache/spark/executor/Executor.scala:
```diff
@@ -277,17 +279,23 @@ private[spark] class Executor(
-    val urls = currentJars.keySet.map { uri =>
+    val urls = userClassPath ++ currentJars.keySet.map { uri =>
       new File(uri.split("/").last).toURI.toURL
-    }.toArray
```
can we keep the toArray here rather then having to do it in 2 places below?

tgravescs commented:
@vanzin  can you detail the testing you did on this.

"The existing "child-first" class loader also had to be fixed. It didn't handle resources, and it
was also doing some things that ended up causing JVM errors depending on how things
were being called."

Can you describe that in more detail, perhaps a couple examples?  Just want to understand this better.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -762,46 +764,37 @@ object Client extends Logging {
     }
-
-    // Append all jar files under the working directory to the classpath.
-    addClasspathEntry(Environment.PWD.$() + Path.SEPARATOR + "*", env)
```
why was * removed? I can't remember what but I thought certain things broke when this was removed.  Maybe that no longer applies though.  You tested the various ways of enabling logging?

vanzin replied:
It's removed for two reasons:

- It didn't serve any practical purpose
- It could potentially lead to behavior that diverged from other cluster managers

All jars distributed with `--jars` are added to the classpath automatically, withouth the need for this. The directory itself is also added, so things like `log4j.properties` uploaded by the user are in the classpath.

The only change this causes is that files and archives (`--files` and `--archives`) would also end up in the app's classpath. This is the part that diverges from other cluster managers - if you use `--files`  to add a jar file in standalone mode, the classes in that jar will not show up in the app's classpath. In Yarn mode they would, and I think that's wrong.

tgravescs replied:
I agree, it would be good to keep consistent.  I just wanted to make sure we didn't break anything by removal.  It sounds like you tested all the things I can think of.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -817,20 +810,13 @@ object Client extends Logging {
    * @param env       Map holding the environment variables.
    */
   private def addFileToClasspath(
-      path: String,
```
need to update header for new param

tgravescs commented on yarn/src/test/resources/log4j.properties:
```diff
@@ -16,7 +16,7 @@
 
 # Set everything to be logged to the file target/unit-tests.log
-log4j.rootCategory=INFO, file
+log4j.rootCategory=DEBUG, file
```
did you mean to leave this debug?

vanzin replied:
Yes. It never made much sense to me to have test logs restricted to `INFO`.

tgravescs commented on yarn/src/test/resources/log4j.properties:
```diff
@@ -25,4 +25,4 @@ log4j.appender.file.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss.SSS} %t %p %c{
 # Ignore messages below warning level from Jetty, because it's a bit verbose
 log4j.logger.org.eclipse.jetty=WARN
-org.eclipse.jetty.LEVEL=WARN
+log4j.logger.org.apache.hadoop=WARN
```
again did you mean to leave this?

tgravescs commented:
took initial pass through.  I need to look at a few parts in more detail and try this out.

Commit:
Some feedback.

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	core/src/main/scala/org/apache/spark/SparkConf.scala

SparkQA commented:
[Test build #25573 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25573/consoleFull) for   PR 3233 at commit [`b6497f9`](https://github.com/apache/spark/commit/b6497f92e3604e748ca72d57a5c98c5dbd6bb5e7).
 * This patch merges cleanly.

vanzin commented:
Hi @tgravescs , thanks for taking a look. Aside from all the unit tests I added, I explained the testing I did, including the code I used, in my very first comment at the top. Did you have any specific questions about that?

SparkQA commented:
[Test build #25573 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25573/consoleFull) for   PR 3233 at commit [`b6497f9`](https://github.com/apache/spark/commit/b6497f92e3604e748ca72d57a5c98c5dbd6bb5e7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25573/
Test PASSed.

tgravescs commented:
Nope your first comment answers it.  Sorry I had read that a while ago but forgot about it.  thanks.

tgravescs commented on docs/configuration.md:
```diff
@@ -257,13 +257,12 @@ Apart from these, the following properties are also available, and may be useful
 </tr>
 <tr>
-  <td><code>spark.files.userClassPathFirst</code></td>
+  <td><code>spark.executor.userClassPathFirst</code></td>
```
we need to document spark.driver.userClassPathFirst

vanzin replied:
I was gonna do that but then I noticed none of the other options (e.g. `spark.driver.extraClassPath`) are actually documented anywhere... since they're all pretty much the same (only the context changes), I could reorganize the documentation a bit so that there's this section that lists options that exist for driver and executors, and then add this one, but it's starting to feel like a separate change at this point.

tgravescs replied:
the other ones like driver.extraClassPath have sparkSubmit command line options though. This one doesn't unless I missed it.

andrewor14 replied:
now they are documented, @vanzin could you document this one too?

tgravescs commented:
this is touching code in core so cc @pwendell @JoshRosen

Commit:
Mention new setting in documentation.

SparkQA commented:
[Test build #25842 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25842/consoleFull) for   PR 3233 at commit [`2e6c4b7`](https://github.com/apache/spark/commit/2e6c4b74405576e4f21a37306a32cfe0f0ed6dda).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25842 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25842/consoleFull) for   PR 3233 at commit [`2e6c4b7`](https://github.com/apache/spark/commit/2e6c4b74405576e4f21a37306a32cfe0f0ed6dda).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25842/
Test FAILed.

vanzin commented:
Are these tests flaky? I've seen them failing on an off in different PRs...

vanzin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #25849 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25849/consoleFull) for   PR 3233 at commit [`2e6c4b7`](https://github.com/apache/spark/commit/2e6c4b74405576e4f21a37306a32cfe0f0ed6dda).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25849 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25849/consoleFull) for   PR 3233 at commit [`2e6c4b7`](https://github.com/apache/spark/commit/2e6c4b74405576e4f21a37306a32cfe0f0ed6dda).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            logInfo("Adding " + url + " to class loader")`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25849/
Test PASSed.

MattWhelan commented on core/src/main/scala/org/apache/spark/executor/ExecutorURLClassLoader.scala:
```diff
@@ -32,36 +35,40 @@ private[spark] trait MutableURLClassLoader extends ClassLoader {
-  override def findClass(name: String): Class[_] = {
+  override def loadClass(name: String, resolve: Boolean): Class[_] = {
     try {
-      userClassLoader.findClass(name)
```
You're bypassing the cache here.  The normal order is cache, parent, self.  You're doing self, cache, parent, self.  I think you probably want cache, self, parent?

The problem with bypassing cache is that findClass doesn't seem to protect against defining the same class twice, using the same bytes, which throws a LinkageError.

Also, you need synchronization here.  ClassLoader.loadClass(String, boolean) synchronizes on 'this' in 1.6, and on a per-name lock in 1.7.  The former is prone to deadlocks in certain cases.

vanzin replied:
What do you mean by "cache"? There's no reference to "cache" anywhere in the documentation of `ClassLoader`. `loadClass` is only overridden so that the normal lookup order is inverted - normally it would be `parent, this` and now it's `this, parent`. If there's any caching done anywhere, I'd expect the actual class loader implementation (`super` or `parentClassLoader` in this code) to handle that.

As for locking, the comments from the JDK 1.7 docs seems to indicate the Spark use case is safe (it mentions deadlocks in the cases of non-hierarchical classloader, which I don't think applies here). But, if that's a concern, adding per-name locking shouldn't be hard (but could be pretty expensive depending on how it's done).

vanzin replied:
BTW, based on your comment (the link to the IBM article) on the other PR, it seems it would be safer to make this method synchronized just to be safe. I'd be surprised if any code in Spark would trigger a deadlock (it would take some really sketchy code in a class's static initializer to trigger that), but better to be safe.

Commit:
Add locking in loadClass() to avoid deadlocks.

See discussions in SPARK-5358 and related PRs.

SparkQA commented:
[Test build #25976 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25976/consoleFull) for   PR 3233 at commit [`fbb8ab5`](https://github.com/apache/spark/commit/fbb8ab521eb734ca2727d7620e28dfdcaab93a7f).
 * This patch merges cleanly.

MattWhelan commented:
Sorry, "cache" was vague.  It's often called that, but what it really refers to is the call to findLoadedClass(name) near the top of the loadClass method in ClassLoader.  That call looks up the name in this CL's cache.  The actual implementation is a native method in the JVM, not some sort of data structure it ClassLoader itself.  

Since you're calling findClass before findLoadedClass, you're going to defineClass on the same class more than once, in some cases, which throws that LinkageError

vanzin commented:
@MattWhelan actually I'm not calling `findClass` at all, exactly because I ran into the linkage error with the original code. So my change is pretty much the same thing you're doing in #4166, except I took a finer-grained locking approach à la Java 7.

Commit:
Call the right loadClass() method on the parent.

MattWhelan commented:
@vanzin Ok, but you were when I initially made that comment :)

Now, you're relying on the fact that your own parent classloader is null.  So it's cache, boot*, self, parent now.  Which should work.

* boot being the boot classloader, which is the default if parent is null.

SparkQA commented:
[Test build #25980 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25980/consoleFull) for   PR 3233 at commit [`3cb6498`](https://github.com/apache/spark/commit/3cb649826d99fa6c895582b20808d36d41e39ae5).
 * This patch merges cleanly.

vanzin commented:
>  but you were when I initially made that comment :)

Not to be nitpicky, but you commented on a line that I was removing.

SparkQA commented:
[Test build #25976 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25976/consoleFull) for   PR 3233 at commit [`fbb8ab5`](https://github.com/apache/spark/commit/fbb8ab521eb734ca2727d7620e28dfdcaab93a7f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            logInfo("Adding " + url + " to class loader")`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25976/
Test PASSed.

SparkQA commented:
[Test build #25980 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25980/consoleFull) for   PR 3233 at commit [`3cb6498`](https://github.com/apache/spark/commit/3cb649826d99fa6c895582b20808d36d41e39ae5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            logInfo("Adding " + url + " to class loader")`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25980/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	core/src/main/scala/org/apache/spark/SparkConf.scala
	core/src/main/scala/org/apache/spark/util/Utils.scala
	core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala

sryza commented:
The new spark.driver.userClassPathFirst property seems a little strange to me in that, IIUC, it only takes effect when the driver is started through the application master.  Is my understanding correct?  Are there use cases you've come across where you think it's necessary?  If not, it might be better to leave it out.

vanzin commented:
> The new spark.driver.userClassPathFirst property seems a little strange to me in that, IIUC, it only takes effect when the driver is started through the application master.

That's incorrect. It applies both in client and cluster mode - thus the changes in `SparkSubmit.scala`. The use case for it is when a user app wants to use a different version of some library that is in the Spark assembly; for example, this could allow you to use your own version of Jetty without Spark having to shade/relocate it.

SparkQA commented:
[Test build #26271 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26271/consoleFull) for   PR 3233 at commit [`25d4fed`](https://github.com/apache/spark/commit/25d4fed5510575672586956808a52fef78d100e9).
 * This patch merges cleanly.

sryza commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -375,4 +390,64 @@ private[spark] object SparkConf {
+      throw new IllegalArgumentException("Need new config name or deprecation message.")
+    }
+
+    def warn() = {
```
: Unit =

SparkQA commented:
[Test build #26271 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26271/consoleFull) for   PR 3233 at commit [`25d4fed`](https://github.com/apache/spark/commit/25d4fed5510575672586956808a52fef78d100e9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            logInfo("Adding " + url + " to class loader")`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26271/
Test PASSed.

sryza commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -350,9 +351,20 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
+        "1.3"),
+      DeprecatedConfig("spark.yarn.user.classpath.first", null, "1.3",
+        "Use spark.{driver,executor}.userClassPathFirst instead."))
+    configs.map { x => (x.oldName, x) }.toMap
```
Nit: use parentheses instead of curly braces here

vanzin replied:
It seems like both patterns are used in Spark. I kinda prefer the braces, makes it clearer (to me) that it's a closure.

andrewor14 replied:
yeah I actually prefer this style

sryza commented on core/src/main/scala/org/apache/spark/executor/ExecutorURLClassLoader.scala:
```diff
@@ -32,36 +36,52 @@ private[spark] trait MutableURLClassLoader extends ClassLoader {
 }
 
 private[spark] class ChildExecutorURLClassLoader(urls: Array[URL], parent: ClassLoader)
```
Can we add some documentation above all these ClassLoader classes describing what their differences are and why they exist?

sryza commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -154,20 +158,69 @@ private[spark] object CoarseGrainedExecutorBackend extends Logging {
+      |   --hostname <hostname>
+      |   --cores <cores>
+      |   --app-id <appid>
+      |   [--worker-url <workerUrl>]
```
I would take these square brackets out - other Spark usage commands with optional params don't have them.

sryza commented on core/src/main/scala/org/apache/spark/executor/ExecutorURLClassLoader.scala:
```diff
@@ -32,36 +36,52 @@ private[spark] trait MutableURLClassLoader extends ClassLoader {
-  }
 }
 
 private[spark] class ExecutorURLClassLoader(urls: Array[URL], parent: ClassLoader)
```
This shouldn't be called ExecutorURLClassLoader if it's being used to load things in the driver.

sryza commented on docs/configuration.md:
```diff
@@ -285,13 +285,13 @@ Apart from these, the following properties are also available, and may be useful
 </tr>
 <tr>
-  <td><code>spark.files.userClassPathFirst</code></td>
+  <td><code>spark.driver.userClassPathFirst, spark.executor.userClassPathFirst</code></td>
```
These should be split up into two different config properties.

vanzin replied:
It would kinda suck to duplicate the actual description, since they'd be identical.

andrewor14 replied:
I agree with @sryza. If we're worried about duplicating description we can always say `Same as spark.driver.userClassPathFirst, but for the executors` or something.

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -760,7 +756,13 @@ object Client extends Logging {
    * Populate the classpath entry in the given environment map.
-   * This includes the user jar, Spark jar, and any extra application jars.
+   *
+   * User jars are generally not added to the JVM's classpath; those are handled by the AM and
```
Are we going to extra pains to support spark.yarn.user.classpath.first here?  If so, I think we should just drop support for it, because it was never documented in the first place.

vanzin replied:
There's very little code needed to keep this option working; I'd rather leave it around as deprecated for one more release, since even though it's not documented, it's been recommended to people to fix certain issues before.

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -451,8 +452,20 @@ private[spark] class ApplicationMaster(
+        new ExecutorURLClassLoader(urls, Utils.getContextOrSparkClassLoader)
+      }
+
+    val mainMethod = userClassLoader.loadClass(args.userClass)
```
What's the difference between using loader.loaderClass as we do here vs. Class.forName elsewhere?  Can we keep them consistent?

vanzin replied:
`loader.loadClass(foo)` should be equivalent to `Class.forName(foo, true, loader)` but IMO looks better. Where do you see calls to the latter?

`Class.forName(String)`, which is used elsewhere in this class, has different semantics.

sryza commented on core/src/main/scala/org/apache/spark/executor/ExecutorURLClassLoader.scala:
```diff
@@ -32,36 +36,52 @@ private[spark] trait MutableURLClassLoader extends ClassLoader {
+  extends URLClassLoader(urls, null) with MutableURLClassLoader {
+
+  private val parentClassLoader = new ParentClassLoader(parent)
+  private val locks = new ConcurrentHashMap[String, Object]()
```
Can we add some doc on why these locks are needed?

sryza commented:
This mostly looks good to me other than my inline comments.  It would be good if someone could take a look at the Spark Standalone changes, because I don't understand that part of the code very well.

Commit:
Review feedback.

Commit:
Move class loaders around and make names more meaninful.

They're not used just by executors, so it makes sense to put them
in a more generic package. Also simplify things a bit by getting
rid of the "MutableURLClassLoader" trait.

SparkQA commented:
[Test build #26377 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26377/consoleFull) for   PR 3233 at commit [`0e6ef19`](https://github.com/apache/spark/commit/0e6ef196f3a2f5bffd3f710f57e72390b502030c).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26377 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26377/consoleFull) for   PR 3233 at commit [`0e6ef19`](https://github.com/apache/spark/commit/0e6ef196f3a2f5bffd3f710f57e72390b502030c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26377/
Test PASSed.

lianhuiwang commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -327,8 +326,14 @@ object SparkSubmit {
-      Thread.currentThread.getContextClassLoader)
+    val loader =
+      if (sysProps.getOrElse("spark.driver.userClassPathFirst", "false").toBoolean) {
+        new ChildFirstURLClassLoader(new Array[URL](0),
```
i think we just use ChildExecutorURLClassLoader and ExecutorURLClassLoader that it has existed. so we reduce code to change.

vanzin replied:
That kinda clashes with Sandy's comment, which I agree with, that calling them "Executor" is misleading since they're used in other contexts. And frankly, compared to the other changes here, changing those class names is a very small one.

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
	yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala
	yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala
	yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala
	yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala
	yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala

Commit:
Fix pyspark/yarn-cluster test.

Make it go through spark-submit like other tests. Also clean up some other
things in the process.

SparkQA commented:
[Test build #26694 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26694/consoleFull) for   PR 3233 at commit [`70d4044`](https://github.com/apache/spark/commit/70d40444911a11d72290a8c669aeb3a69d3afb47).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26694 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26694/consoleFull) for   PR 3233 at commit [`70d4044`](https://github.com/apache/spark/commit/70d40444911a11d72290a8c669aeb3a69d3afb47).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26694/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

SparkQA commented:
[Test build #26774 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26774/consoleFull) for   PR 3233 at commit [`0e6d6be`](https://github.com/apache/spark/commit/0e6d6be6c6f0c912ef0eced95f79dd5f006b69dd).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26774 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26774/consoleFull) for   PR 3233 at commit [`0e6d6be`](https://github.com/apache/spark/commit/0e6d6be6c6f0c912ef0eced95f79dd5f006b69dd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26774/
Test FAILed.

vanzin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #26863 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26863/consoleFull) for   PR 3233 at commit [`0e6d6be`](https://github.com/apache/spark/commit/0e6d6be6c6f0c912ef0eced95f79dd5f006b69dd).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26863 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26863/consoleFull) for   PR 3233 at commit [`0e6d6be`](https://github.com/apache/spark/commit/0e6d6be6c6f0c912ef0eced95f79dd5f006b69dd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26863/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
	core/src/main/scala/org/apache/spark/executor/ExecutorURLClassLoader.scala
	yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala
	yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala

SparkQA commented:
[Test build #26932 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26932/consoleFull) for   PR 3233 at commit [`3f768e3`](https://github.com/apache/spark/commit/3f768e31e9d454522c6bb71be90259fadf4a7071).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26932 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26932/consoleFull) for   PR 3233 at commit [`3f768e3`](https://github.com/apache/spark/commit/3f768e31e9d454522c6bb71be90259fadf4a7071).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26932/
Test FAILed.

Commit:
Merge branch 'master' into SPARK-2996

SparkQA commented:
[Test build #26940 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26940/consoleFull) for   PR 3233 at commit [`a1b8d7e`](https://github.com/apache/spark/commit/a1b8d7e949e30a76d5fe42d0eb1095effe718d10).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26940 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26940/consoleFull) for   PR 3233 at commit [`a1b8d7e`](https://github.com/apache/spark/commit/a1b8d7e949e30a76d5fe42d0eb1095effe718d10).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26940/
Test FAILed.

vanzin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #26953 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26953/consoleFull) for   PR 3233 at commit [`a1b8d7e`](https://github.com/apache/spark/commit/a1b8d7e949e30a76d5fe42d0eb1095effe718d10).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26953 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26953/consoleFull) for   PR 3233 at commit [`a1b8d7e`](https://github.com/apache/spark/commit/a1b8d7e949e30a76d5fe42d0eb1095effe718d10).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26953/
Test PASSed.

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -380,4 +392,64 @@ private[spark] object SparkConf {
+   * deprecated keys.
+   *
+   * @param userKey Configuration key from the user / caller.
+   * @param warn Whether to print a warning if the key is depreated. Warnings will be printed
```
deprecated

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -380,4 +392,64 @@ private[spark] object SparkConf {
+
+  /**
+   * Returns the configuration key to use for the given user-provided key, translating
+   * deprecated keys.
```
How about
```
Translate the configuration key if it is deprecated, or just return the provided key.
```

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -380,4 +392,64 @@ private[spark] object SparkConf {
+      if (deprecatedKey.newName != null) deprecatedKey.newName else userKey
+    } else {
+      userKey
+    }
```
you can rewrite this without the null checks
```
val deprecatedKey = deprecatedConfigs.get(userKey)
if (warn) {
  deprecatedKey.foreach(_.warn())
}
deprecatedKey
  .flatMap { k => Option(k.newName) }
  .getOrElse(userKey)
```
up to you

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -351,9 +352,20 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
-private[spark] object SparkConf {
+private[spark] object SparkConf extends Logging {
+
+  private val deprecatedConfigs = {
```
can you add a type to this

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -139,7 +140,7 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
   /** Set a parameter if it isn't already configured */
   def setIfMissing(key: String, value: String): SparkConf = {
-    settings.putIfAbsent(key, value)
+    settings.putIfAbsent(getConfKey(key, true), value)
```
can you do `warn = true` here? also in L71

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/Executor.scala:
```diff
@@ -288,17 +291,23 @@ private[spark] class Executor(
-    userClassPathFirst match {
-      case true => new ChildExecutorURLClassLoader(urls, currentLoader)
-      case false => new ExecutorURLClassLoader(urls, currentLoader)
+    }).toArray
```
would be easier to read if it's `userClassPath.toArray ++ ...`

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/Executor.scala:
```diff
@@ -344,18 +353,20 @@ private[spark] class Executor(
-        if (!urlClassLoader.getURLs.contains(url)) {
-          logInfo("Adding " + url + " to class loader")
-          urlClassLoader.addURL(url)
+        if (currentJars.get(name).orElse(currentJars.get(localName)).getOrElse(-1L) < timestamp) {
```
can you put this in a variable, something like
```
val existingTimestamp = currentJars.get(name)
  .orElse(currentJars.get(localName))
  .getOrElse(-1L)
if (existingTimestamp < timestamp) {
  ...
}
```
Also, why do we check for the `localName` here but not before?

vanzin replied:
Probably because I didn't test files explicitly, but it's probably safer to do it also for them.

vanzin replied:
Actually, it's because, unlike the initial list of jars distributed by yarn (which is now explicitly provided in the command line), the executor doesn't know about the initial list of files. So doing the same thing would require implementing that.

Let me try to remember why exactly I had to add this in the first place. It's been a while.

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/MesosExecutorBackend.scala:
```diff
@@ -73,7 +73,8 @@ private[spark] class MesosExecutorBackend
       slaveInfo.getHostname,
-      env)
+      env,
+      Nil)
```
no need for this since it already defaults to Nil

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/local/LocalBackend.scala:
```diff
@@ -56,7 +56,7 @@ private[spark] class LocalActor(
 
   private val executor = new Executor(
-    localExecutorId, localExecutorHostname, SparkEnv.get, isLocal = true)
+    localExecutorId, localExecutorHostname, SparkEnv.get, Nil, isLocal = true)
```
same here

andrewor14 commented on core/src/main/scala/org/apache/spark/util/MutableURLClassLoader.scala:
```diff
@@ -0,0 +1,103 @@
+   */
+  private val locks = new ConcurrentHashMap[String, Object]()
+
+  override def loadClass(name: String, resolve: Boolean): Class[_] = {
```
should we keep this protected?

andrewor14 replied:
also, should we override `findClass` instead? Quoting the docs from `ClassLoader.java`
```
  ...
  * <p> Subclasses of <tt>ClassLoader</tt> are encouraged to override {@link
  * #findClass(String)}, rather than this method.  </p>
  ...
  protected Class<?> loadClass(String name, boolean resolve)
  ...  
```

vanzin replied:
`findClass` doesn't work because of two things:

- it's protected, so you can't call `parentClassLoader.findClass` since you'll get a compilation error
- you can't rely on the super class's implementation to call `findClass` on the parent for you, since that's not in the specification the method. In fact, `ClassLoader.findClass()` just throws.

And you can't call `parentClassLoader.loadClass()` from `findClass()`, since you'll end up with linkage errors.

andrewor14 replied:
ok, for this patch that's fine since we only ever call `findClass` from `loadClass` anyway (I believe this is the case). Since we're overriding load class then nobody will call findClass because it's protected. Well, unless someone goes ahead and extends this class, but I think it's fine to assume that nobody will.

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -453,12 +453,24 @@ private[spark] class ApplicationMaster(
+        new ChildFirstURLClassLoader(urls, Utils.getContextOrSparkClassLoader)
+      } else {
+        new MutableURLClassLoader(urls, Utils.getContextOrSparkClassLoader)
+      }
```
can we add a short comment that explains this is doing the user class path first thing

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -453,12 +453,24 @@ private[spark] class ApplicationMaster(
+      new URL("file:" + new File(entry.getPath()).getAbsolutePath())
+    }
+    val userClassLoader =
+      if (Client.isUserClassPathFirst(sparkConf, true)) {
```
can you do `isDriver = true`

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -962,4 +952,23 @@ object Client extends Logging {
+  /**
+   * Joins all the path components using Path.SEPARATOR.
+   */
+  def buildPath(components: String*) = {
```
please add return type

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -794,48 +803,37 @@ object Client extends Logging {
+   * @param conf Spark configuration.
    */
-  private def addUserClasspath(
+  def getUserClasspath(
```
if we're exposing this we should probably add an alias
```
def getUserClassPath(conf: SparkConf): Array[URI] = getUserClassPath(null, conf)
```
so we don't pass null around

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -318,6 +314,13 @@ private[spark] class Client(
+    // of SPARK_HOME.
+    sparkConf.getOption("spark.home").orElse(sys.env.get("SPARK_HOME")).foreach { path =>
+      env("SPARK_HOME") = path
+    }
```
a little confused. Is this related to this patch?

andrewor14 replied:
also prefer breaking the line at `orElse` and `forEach`

vanzin replied:
It's not entirely related to the patch. But without this I just could not run pyspark apps on yarn as part of testing. I couldn't for the life of me find out how SPARK_HOME propagated to executors before.

vanzin replied:
I think I figured this one out.

The assembly generated by maven includes the python files needed to run pyspark. The assembly generated by sbt doesn't. So if you deploy the sbt assembly, since `SPARK_HOME` is not propagated, it won't include the py4j and pyspark modules in `PYTHONPATH`, and pyspark jobs will fail.

In any case, unrelated to this issue and this is probably the wrong fix anyway, so I'll revert this part.

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -239,23 +236,22 @@ private[spark] class Client(
+     * Each resource is represented by a 3-tuple of:
      *   (1) destination resource name,
      *   (2) local path to the resource,
      *   (3) Spark property key to set if the scheme is not local, and
```
remove `, and`

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -846,27 +844,19 @@ object Client extends Logging {
+      addClasspathEntry(uri.getPath, env)
+    } else if (fileName != null) {
+      addClasspathEntry(buildPath(
+        YarnSparkHadoopUtil.expandEnvironment(Environment.PWD), fileName), env)
```
nice clean up. I wonder why the ignoring `URISyntaxException` was there in the first place...

andrewor14 replied:
also, note to self for a future patch: we should probably just break this into two separate methods. Right now it takes 2 arguments that are mutually exclusive so that one of them is always null

vanzin replied:
Actually they're not mutually exclusive. See L817:

        addFileToClasspath(new URI(sparkJar(sparkConf)), SPARK_JAR, env)

You could do special handling at that call site, but then you'd be duplicating the logic to handle "local:" URIs.

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -239,23 +236,22 @@ private[spark] class Client(
      */
     List(
-      (SPARK_JAR, sparkJar(sparkConf), CONF_SPARK_JAR, false),
-      (APP_JAR, args.userJar, CONF_SPARK_USER_JAR, true),
```
what implications does removing setPerms here have?

andrewor14 replied:
is it simply that we want everything to be `APP_FILE_PERMISSION` i.e. "world-wide readable and owner writable -> rw-r--r--" per the comment in L681?

vanzin replied:
The motivation is to simplify code. It doesn't seem like there's any functional requirement to have different files have different permissions.

In fact, the call to set permissions could potentially be removed altogether, since all these files are under a directory managed by Yarn which has 0700 permissions anyway.

andrewor14 replied:
ok, that's what I thought. Let's do the remove set permissions thing separately

andrewor14 commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala:
```diff
@@ -17,23 +17,30 @@
 import org.apache.spark.util.Utils
 
+/**
+ * Integration tests for Yarn; these tests use a mini Yarn cluster to run Spark-on-Yarn
```
I think elsewhere we use YARN instead of Yarn

andrewor14 commented on yarn/src/test/resources/test.resource:
```diff
@@ -0,0 +1 @@
+ORIGINAL
```
do you actually need this file?

andrewor14 commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala:
```diff
@@ -117,99 +116,149 @@ class YarnClusterSuite extends FunSuite with BeforeAndAfterAll with Matchers wit
+    checkResult(result)
+  }
+
+  private def testUseClassPathFirst(clientMode: Boolean) = {
```
these all seem to return `Unit`. It would be good if you could add the return types here

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -782,7 +785,13 @@ object Client extends Logging {
-   * This includes the user jar, Spark jar, and any extra application jars.
+   *
+   * User jars are generally not added to the JVM's classpath; those are handled by the AM and
+   * executor backend. When the deprecated `spark.yarn.user.classpath.first` is used, user jars are
```
It would be clearer if you explicitly say that `those are loaded by the AM or the executor backend dynamically`, because it's technically added to the JVM's class path, just that it's added later.

vanzin replied:
> because these jars are technically added to the JVM's class path

Not true. They're added to a different class loader, and thus do not show up in `sys.props("java.class.path")` at all.

andrewor14 replied:
ah right they're added to the special class loader that is used in the driver / executors.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/DriverWrapper.scala:
```diff
@@ -28,21 +30,31 @@ import org.apache.spark.util.{AkkaUtils, Utils}
 
       case _ =>
-        System.err.println("Usage: DriverWrapper <workerUrl> <driverMainClass> [options]")
+        System.err.println("Usage: DriverWrapper <workerUrl> <userJar> <driverMainClass> [options]")
```
we're technically changing the public API here, which is probably fine since there's not really a way around it in this case. @pwendell thoughts?

vanzin replied:
Is this object supposed to be public in the first place? Seems like it's only used through SparkSubmit (much like Yarn's `Client`, although that one is public only for backwards compatibility).

andrewor14 commented:
@vanzin thanks for doing this. This looks mostly good to me; most of my comments are fairly minor. @tgravescs did you have more comments or does it look good to you?

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

Commit:
Review feedback.

Commit:
Remove 'test.resource' file, create it dynamically.

This uncovered some fun things. To do this, we need to override the
executor's class path to make the test actually test what it intends
to. But that config was being used by the build system to propagate
the test classpath. So change the build (surefire and sbt test runner)
to use SPARK_DIST_CLASSPATH for that, like the maven scalatest plugin
does.

Also, propagate SPARK_DIST_CLASSPATH to the executor's classpath when
running on Yarn.

Commit:
Remove SPARK_HOME propagation.

I ran into issues because the sbt-built assembly does not include pyspark
files. That is a bug but it's unrelated to this patch, so undo the change.

SparkQA commented:
[Test build #27157 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27157/consoleFull) for   PR 3233 at commit [`a1499e2`](https://github.com/apache/spark/commit/a1499e21bcd4f95742848ef369d98eabb9b54b8b).
 * This patch **does not merge cleanly**.

Commit:
Merge branch 'master' into SPARK-2996

Conflicts:
	yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala

SparkQA commented:
[Test build #27159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27159/consoleFull) for   PR 3233 at commit [`9cf9cf1`](https://github.com/apache/spark/commit/9cf9cf1e634116e2efcceb89772a354a01f48e89).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27157 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27157/consoleFull) for   PR 3233 at commit [`a1499e2`](https://github.com/apache/spark/commit/a1499e21bcd4f95742848ef369d98eabb9b54b8b).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27157/
Test FAILed.

SparkQA commented:
[Test build #27159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27159/consoleFull) for   PR 3233 at commit [`9cf9cf1`](https://github.com/apache/spark/commit/9cf9cf1e634116e2efcceb89772a354a01f48e89).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27159/
Test PASSed.

andrewor14 commented:
Ok @vanzin I'm merging this into master and 1.3 thanks a lot for reiterating on the reviews quickly. I will remove the random println that I commented.
